### PR TITLE
W3C input states (17.2)

### PR DESCRIPTION
**Pull request by dpgraham**
(I'm going to merge this and future commits into the one dpgraham-w3c branch until that one is ready to be merged into master)

* Added state models from W3C spec: https://www.w3.org/TR/webdriver/#input-source-state
* Added PointerInputState, KeyInputState and amended the source states to return default state objects

### CODE REVIEW

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/w3c/models/InputSource.java [hunk 1]
```diff
@@ -47,10 +63,32 @@ public void setType(InputSourceType type) {
...
+     * Get the initial state of an Input Source
+     * @return
+     */
+    public InputStateInterface getDefaultState() {
+        switch (getType()) {
+            case POINTER:
+                return new PointerInputState();
+            case KEY:
+                return new KeyInputState();
+            case NONE:
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/w3c/models/InputSource.java]: this entry is not needed or at least the following `return null` may be omitted

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/w3c/models/InputSource.java [hunk 2]
```diff
@@ -47,10 +63,32 @@ public void setType(InputSourceType type) {
...
+        }
+    }
+
+
     @Nullable
     public PointerType getPointerType(){
         if(parameters != null) {
             return parameters.getPointerType();
+        } else if (type == InputSourceType.POINTER) {
+            // NOTE: The spec specifies that the default should be MOUSE. This is an exception
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/w3c/models/InputSource.java]: 👍

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/w3c/models/InputSource.java [hunk 3]
```diff
@@ -47,10 +63,32 @@ public void setType(InputSourceType type) {
...
+                return null;
+            default:
+                return null;
+        }
+    }
+
+
     @Nullable
     public PointerType getPointerType(){
         if(parameters != null) {
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/w3c/models/InputSource.java]: a space is missing

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/w3c/models/InputSource.java [hunk 4]
```diff
@@ -64,14 +102,14 @@ public PointerType getPointerType(){
...
-    public class Action {
-        private ActionType type; // type of action
-        private Long duration; // time in milliseconds
-        private String origin; // origin; could be viewport, pointer or <ELEMENT_ID>
-        private Integer button; // Button that is being pressed. Defaults to 0.
-        private Long x; // x coordinate of pointer
-        private Long y; // y coordinate of pointer
-        private String value; // a string containing a single Unicode code point
+    public static class Action {
+        ActionType type; // type of action
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/w3c/models/InputSource.java]: I'd keep `private`. Codacy does not appreciate package-level access

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/w3c/state/InputStateTable.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+import java.util.Map;
+
+/**
+ * Keep the state of all active input sources
+ *
+ * (defined in 17.1 of spec)
+ */
+public class InputStateTable {
+
+    private Map<String, InputStateInterface> stateTable = new HashMap<>();
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/w3c/state/InputStateTable.java]: `final`

Also, I'd prefer to use ImmutableMapBuilder

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/w3c/state/KeyInputState.java [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+    public KeyInputState() {
+        pressed = new HashSet<>();
+        alt = false;
+        shift = false;
+        ctrl = false;
+        meta = false;
+    }
+
+
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/w3c/state/KeyInputState.java]: extra line break

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/w3c/state/KeyInputState.java [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+public class KeyInputState implements InputStateInterface {
+    private Set<String> pressed;
+    private boolean alt;
+    private boolean shift;
+    private boolean ctrl;
+    private boolean meta;
+
+    public KeyInputState() {
+        pressed = new HashSet<>();
+        alt = false;
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/w3c/state/KeyInputState.java]: the default values can be set on variable definition

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/w3c/state/PointerInputState.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+ * https://www.w3.org/TR/webdriver/#input-source-state
+ */
+public class PointerInputState implements InputStateInterface {
+    private InputSource.PointerType subtype;
+    private Set<Integer> pressed;
+    private int x;
+    private int y;
+
+    public PointerInputState() {
+        pressed = new HashSet<>();
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/w3c/state/PointerInputState.java]: same as above

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/w3c/state/InputStateTable.java (hunk 5 above)
Comment (dpgraham) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/w3c/state/InputStateTable.java]: Do you have an example? I couldn't find ImmutableMap.Builder or ImmutableMapBuilder in the libraries.

**Pull request merged**